direct.c

```diff
@@ -357,6 +357,7 @@ rr_data_t direct_request(void *cdata, rr_data_const_t request) {
 					if (debug)
 						printf("Reconnect before WWW auth\n");
 					close(sd);
+					data[1]->body_len = BODY_VANISHED;
```

Comment:
Thank you @fralken for providing your doubt with a hint on how to resolve it!

Unfortunately, after having tried it (1),
it requires resetting _every header that http_has_body tests_ to detect "something left to read"
Thus I was able to pass, but with an ugly:
```c
data[1]->headers = hlist_del(hlist_del(hlist_del(data[1]->headers, "Content-Length"), "Connection"), "Content-Type");
```
… which is fragile too, making us tightly tied to http_has_body implementation.

Now I'm wondering, from a software theory point of view:
shouldn't the `close(sd)` (playing the role of a kind of big red "abort" button) **clear anything related to the old connection** to make sure no artifact of this old connection gets reused by the new connection?
… Of course this full cleanup introduces the complexity that all data of the old connection needed to establish the new one (nonce) has to be secured until having been "consumed" :-(

Thus, unless we are sure no bit of the old connection is necessary after the close() (and then we can do a full clear without the complexity I fear),
I'm tempted to think my BODY_VANISHED was a "more simple" and honest way for the closer code to tell to the consumer code: "whoops, you mean there was a body? I probably thrashed it without noticing, I just hope there was nothing important for you there because now you'd better do as if it never existed".

An alternative would be to delay the close until after the `http_has_body`
(but then how would `direct_request()` be _sure_ the sd it lets going is garbage collected?).

##### Notes

**(1)** As I didn't find which one of my company's always changing proxies behaved this way, I did a quick simulation with nginx:
```nginx
location /401/
    {
        keepalive_timeout 0; # https://stackoverflow.com/a/17010681
        add_header WWW-Authenticate NTLM always;
        return 401; # nginx defaults accompagny the 401 with a pretty message and its Content-Length.
    }
```
Then starting cntlm on port 8080 with a NoProxy * and `curl -x localhost:8080 http://localhost/401/`.
_Of course this setup is nonfunctional_ because this NTLM exchange is rotten (we emit a WWW-Authenticate without implementing the then expected challenge, so we get a "502 Web server reply missing NTLM challenge")\
but at least _it allows reaching the bug_ (cntlm hangs without reaching the 502).

Reply:
Thank you @outtersg for testing my suggestion

> Unfortunately, after having tried it (1),
it requires resetting every header that http_has_body tests to detect "something left to read"

I was aware of the logic of `http_has_body` but not completely sure. That's why I gave you this alternative (of setting the content length to 0):

```
data[1]->headers = hlist_mod(data[1]->headers, "Content-Length", "0", 1);
```
Can you do a test with this (modify the header instead of deleting it)

> Thus I was able to pass, but with an ugly:

You can write it in a less ugly way:

```
data[1]->headers = hlist_del(data[1]->headers, "Content-Length");
data[1]->headers = hlist_del(data[1]->headers, "Connection");
data[1]->headers = hlist_del(data[1]->headers, "Content-Type");
```
I agree that this is not the best approach anyway.

> Now I'm wondering, from a software theory point of view:
shouldn't the close(sd) (playing the role of a kind of big red "abort" button) clear anything related to the old connection to make sure no artifact of this old connection gets reused by the new connection?

Here the problem is that the remote server closes the connection after a 401 while it should not (we are in the middle of a NTLM authentication sequence). The socket is (going to be) closed but we need to keep it open to go on with the authentication sequence. That's why we close the handler with `close(sd)` and we open a new one. Unfortunately in this way we lose the existing response body (that we don't need anyway).
To manage this special case while keeping the logic unaltered in my opinion the easiest way is to "fake" the response body length to zero.

Thanks for sharing your test scenario, if I have time I'll try it myself but first I ask you if you can test with this line
```
data[1]->headers = hlist_mod(data[1]->headers, "Content-Length", "0", 1);
```

Reply:
My bad 😞 While bla-bla-commenting over `http_has_body()` I didn't even realize that the condition in there distinguished "Content-Length: 0" from no Content-Length, so I just tested your first proposal without trying the second one. Oops.

… And so a quick test confirms that **it works** when overriding Content-Length to 0!

Now a bit of code cleanup and I'll resubmit it (unless you prefer to commit it by yourself? After all, you're the one who resolved it)

Reply:
Good news! Just to be very very precise I ask you to remove the reference to this issue "(see #85)" from the code comment because the source code lives beyond this github repo.

It's ok that **you** submit this commit since you found the issue, we together shared a solution and you tested it. So you deserve it :).